README.md
pySMART.smartx
========================

It's a fork of pySMART with lots of bug fix and enhances.


You can follow the [original-readme](./README_ORIGINAL.rst) to read the doc and checkout the way to use it.

More doc about our fixes and enhance may be add if we have more free time.

Pull request is very welcome!

# Feature
+ Python2/Python3 compatible.
+ Bug fixes.
+ More device-type supports.
  + Output parse intel ssd
  + Some seagate hdd's self_test support
+ Health status API.
+ Support for MacOS

# Quick start

just run `pip install pysmart.smartx` to install this package.

# Changelog
[Follow this link](./changelog.md)


CHANGES.txt
0.3 - 2015/06/10
--------------------------------------------------------------------------------
[BUG FIXES]
    * Now able to parse used endurance indicator (ie: SSD life left) from HGST
      and Seagate SAS SSDs.
    * Fixed crash when attempting to parse self-test entries from HGST and
      Seagate SAS devices with an empty self-test log.
[ENHANCEMENTS]
    * Generally improved detection of empty self-test logs prior to attempting
      to parse self-test entries.
[ADMINISTRATIVE]
    * Added support for FreeBSD with smartctl 6.1+ to the version check in
      utils.py by request. No testing has been performed with FreeBSD, this
      only prevents the code from aborting due to the unknown OS string.

0.2 - 2015/01/06
--------------------------------------------------------------------------------
[BUG FIXES]
    * Now properly detects whether a SMART self-test was initiated when
      requested on SAS/SCSI devices.
    * Now properly detects when a SMART self-test has completed on
      SAS/SCSI devices.
    * The values of diagnostic attributes 'Load_Cycle_Pct_Left' and
      'Start_Stop_Pct_Left' are now set to '-' if the manufacturer's
      specification is '0' (unlimited - typical of SSDs).
    * Diagnostic attribute 'Load_Cycle_Pct_Left' now obtains the intended
      default value '-' when not supported by a device.
[ENHANCEMENTS]
    * Disambiguation of SAS from SCSI is more robust for older devices which
      do not support the SAS PHY log command.
    * Extraction of power-on time from SAS/SCSI devices is more robust.
    * The calculated values of diagnostic attributes 'Load_Cycle_Pct_Left' and
      'Start_Stop_Pct_Left' now include the '%' sign.
[ADMINISTRATIVE]
    * Corrected the package manifest to include the GPLv2 license file.

0.1 - 2014/12/24
--------------------------------------------------------------------------------
    * Initial release.

pySMART/test_entry.py
#Copyright (C) 2014 Marc Herndon
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License,
#version 2, as published by the Free Software Foundation.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
"""
This module contains the definition of the `Test_Entry` class, used to
represent individual entries in a `Device`'s SMART Self-test Log.
"""
class Test_Entry(object):
    """
    Contains all of the information associated with a single SMART Self-test
    log entry. This data is intended to exactly mirror that obtained through
    smartctl.
    """
    def __init__(self, format, num, test_type, status, hours, LBA, remain=None,
                 segment=None, sense=None, ASC=None, ASCQ=None):
        self._format = format
        """
        **(str):** Indicates whether this entry was taken from an 'ata' or
        'scsi' self-test log. Used to display the content properly.
        """
        self.num = num
        """
        **(str):** Entry's position in the log from 1 (most recent) to 21
        (least recent).  ATA logs save the last 21 entries while SCSI logs
        only save the last 20.
        """
        self.type = test_type
        """
        **(str):** Type of test run.  Generally short, long (extended), or
        conveyance, plus offline (background) or captive (foreground).
        """
        self.status = status
        """
        **(str):** Self-test's status message, for example 'Completed without
        error' or 'Completed: read failure'.
        """
        self.hours = hours
        """
        **(str):** The device's power-on hours at the time the self-test
        was initiated.
        """
        self.LBA = LBA
        """
        **(str):** Indicates the first LBA at which an error was encountered
        during this self-test. Presented as a decimal value for ATA/SATA devices
        and in hexadecimal notation for SAS/SCSI devices.
        """
        self.remain = remain
        """
        **(str):** Percentage value indicating how much of the self-test is
        left to perform. '00%' indicates a complete test, while any other
        value could indicate a test in progress or one that failed prior to
        completion. Only reported by ATA devices.
        """
        self.segment = segment
        """
        **(str):** A manufacturer-specific self-test segment number reported
        by SCSI devices on self-test failure. Set to '-' otherwise.
        """
        self.sense = sense
        """
        **(str):** SCSI sense key reported on self-test failure. Set to '-'
        otherwise.
        """
        self.ASC = ASC
        """
        **(str):** SCSI 'Additonal Sense Code' reported on self-test failure.
        Set to '-' otherwise.
        """
        self.ASCQ = ASCQ
        """
        **(str):** SCSI 'Additonal Sense Code Quaifier' reported on self-test
        failure. Set to '-' otherwise.
        """

    def __repr__(self):
        """Define a basic representation of the class object."""
        return "<SMART Self-test [%s|%s] hrs:%s LBA:%s>" % (
            self.type, self.status, self.hours, self.LBA)

    def __str__(self):
        """
        Define a formatted string representation of the object's content.
        Looks nearly identical to the output of smartctl, without overflowing
        80-character lines.
        """
        if self._format == 'ata':
            return "{0:>2} {1:17}{2:30}{3:5}{4:7}{5:17}".format(
                self.num, self.type, self.status, self.remain, self.hours,
                self.LBA)
        else:
            # 'Segment' could not be fit on the 80-char line. It's of limited
            # utility anyway due to it's manufacturer-proprietary nature...
            return ("{0:>2} {1:17}{2:23}{3:7}{4:14}[{5:4}{6:5}{7:4}]".format(
                        self.num, self.type, self.status, self.hours,
                        self.LBA, self.sense, self.ASC, self.ASCQ))

__all__ = ['Test_Entry']


pySMART/test/conftest.py
import pytest


@pytest.fixture
def cmd_scan_open_as_root():
    """
    run `smartctl --scan-open` as root user.
    """
    stderr = ""
    stdout = """
/dev/sda -d sat # /dev/sda [SAT], ATA device
/dev/sdb -d sat # /dev/sdb [SAT], ATA device
/dev/sdc -d sat # /dev/sdc [SAT], ATA device
/dev/sdd -d sat # /dev/sdd [SAT], ATA device
/dev/sde -d sat # /dev/sde [SAT], ATA device
/dev/sdf -d sat # /dev/sdf [SAT], ATA device
/dev/sdg -d scsi # /dev/sdg, SCSI device
/dev/sdh -d scsi # /dev/sdh, SCSI device
/dev/bus/0 -d sat+megaraid,8 # /dev/bus/0 [megaraid_disk_08] [SAT], ATA device
/dev/bus/0 -d sat+megaraid,9 # /dev/bus/0 [megaraid_disk_09] [SAT], ATA device
/dev/bus/0 -d sat+megaraid,10 # /dev/bus/0 [megaraid_disk_10] [SAT], ATA device
/dev/bus/0 -d sat+megaraid,11 # /dev/bus/0 [megaraid_disk_11] [SAT], ATA device
/dev/bus/0 -d sat+megaraid,12 # /dev/bus/0 [megaraid_disk_12] [SAT], ATA device
/dev/bus/0 -d sat+megaraid,13 # /dev/bus/0 [megaraid_disk_13] [SAT], ATA device
/dev/bus/0 -d megaraid,16 # /dev/bus/0 [megaraid_disk_16], SCSI device
/dev/bus/0 -d megaraid,17 # /dev/bus/0 [megaraid_disk_17], SCSI device
    """
    return stdout, stderr


@pytest.fixture
def cmd_scan_open_non_root():
    stderr = ""
    stdout = """
# /dev/sda -d scsi # /dev/sda, SCSI device open failed: Permission denied
# /dev/sdb -d scsi # /dev/sdb, SCSI device open failed: Permission denied
# /dev/sdc -d scsi # /dev/sdc, SCSI device open failed: Permission denied
# /dev/sdd -d scsi # /dev/sdd, SCSI device open failed: Permission denied
# /dev/sde -d scsi # /dev/sde, SCSI device open failed: Permission denied
# /dev/sdf -d scsi # /dev/sdf, SCSI device open failed: Permission denied
# /dev/sdg -d scsi # /dev/sdg, SCSI device open failed: Permission denied
# /dev/sdh -d scsi # /dev/sdh, SCSI device open failed: Permission denied
    """
    return stdout, stderr

pySMART/test/test_macos.py
from contextlib import nested

import mock
import pytest


@pytest.fixture
def macos_cmd_all():
    """
    smartctl -a /dev/disk1
    """
    stderr = ""
    stdout = """
smartctl 6.5 2016-05-07 r4318 [Darwin 16.4.0 x86_64] (local build)
Copyright (C) 2002-16, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF INFORMATION SECTION ===
Model Family:     Apple SD/SM/TS...E/F SSDs
Device Model:     APPLE SSD SM0512G
Serial Number:    S29ANYAG123456
LU WWN Device Id: 5 001234 900000000
Firmware Version: BXW1SA0Q
User Capacity:    500,277,790,720 bytes [500 GB]
Sector Sizes:     512 bytes logical, 4096 bytes physical
Rotation Rate:    Solid State Device
Device is:        In smartctl database [for details use: -P show]
ATA Version is:   ATA8-ACS T13/1699-D revision 4c
SATA Version is:  SATA 3.0, 6.0 Gb/s (current: 6.0 Gb/s)
Local Time is:    Tue Feb  7 14:10:04 2017 CET
SMART support is: Available - device has SMART capability.
SMART support is: Enabled

=== START OF READ SMART DATA SECTION ===
SMART overall-health self-assessment test result: PASSED

General SMART Values:
Offline data collection status:  (0x00)	Offline data collection activity
					was never started.
					Auto Offline Data Collection: Disabled.
Self-test execution status:      (   0)	The previous self-test routine completed
					without error or no self-test has ever
					been run.
Total time to complete Offline
data collection: 		(    0) seconds.
Offline data collection
capabilities: 			 (0x53) SMART execute Offline immediate.
					Auto Offline data collection on/off support.
					Suspend Offline collection upon new
					command.
					No Offline surface scan supported.
					Self-test supported.
					No Conveyance Self-test supported.
					Selective Self-test supported.
SMART capabilities:            (0x0003)	Saves SMART data before entering
					power-saving mode.
					Supports SMART auto save timer.
Error logging capability:        (0x01)	Error logging supported.
					General Purpose Logging supported.
Short self-test routine
recommended polling time: 	 (   2) minutes.
Extended self-test routine
recommended polling time: 	 (  10) minutes.

SMART Attributes Data Structure revision number: 1
Vendor Specific SMART Attributes with Thresholds:
ID# ATTRIBUTE_NAME          FLAG     VALUE WORST THRESH TYPE      UPDATED  WHEN_FAILED RAW_VALUE
  1 Raw_Read_Error_Rate     0x001a   200   200   000    Old_age   Always       -       0
  5 Reallocated_Sector_Ct   0x0033   100   100   000    Pre-fail  Always       -       0
  9 Power_On_Hours          0x0032   099   099   000    Old_age   Always       -       3818
 12 Power_Cycle_Count       0x0032   089   089   000    Old_age   Always       -       10461
169 Unknown_Attribute       0x0013   212   212   010    Pre-fail  Always       -       4527071692672
173 Wear_Leveling_Count     0x0032   195   195   100    Old_age   Always       -       8599765028
174 Host_Reads_MiB          0x0022   099   099   000    Old_age   Always       -       12752293
175 Host_Writes_MiB         0x0022   099   099   000    Old_age   Always       -       11997087
192 Power-Off_Retract_Count 0x0012   099   099   000    Old_age   Always       -       47
194 Temperature_Celsius     0x0022   070   017   000    Old_age   Always       -       30 (Min/Max 13/84)
197 Current_Pending_Sector  0x0022   100   100   000    Old_age   Always       -       0
199 UDMA_CRC_Error_Count    0x001a   200   199   000    Old_age   Always       -       0

SMART Error Log Version: 1
No Errors Logged

SMART Self-test log structure revision number 1
No self-tests have been logged.  [To run self-tests, use: smartctl -t]

Warning! SMART Selective Self-Test Log Structure error: invalid SMART checksum.
SMART Selective self-test log data structure revision number 1
 SPAN  MIN_LBA  MAX_LBA  CURRENT_TEST_STATUS
    1        0        0  Not_testing
    2        0        0  Not_testing
    3        0        0  Not_testing
    4        0        0  Not_testing
    5        0        0  Not_testing
  255        0    65535  Read_scanning was never started
Selective self-test flags (0x0):
  After scanning selected spans, do NOT read-scan remainder of disk.
If Selective self-test is pending on power-up, resume after 0 minute delay.

    """
    return stdout, stderr


@pytest.fixture
def macos_cmd_scan_open_as_root():
    stderr = ""
    stdout = """
IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/RP06@1C,5/IOPP/SSD0@0/AppleAHCI/PRT0@0/IOAHCIDevice@0/AppleAHCIDiskDriver/IOAHCIBlockStorageDevice -d ata # IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/RP06@1C,5/IOPP/SSD0@0/AppleAHCI/PRT0@0/IOAHCIDevice@0/AppleAHCIDiskDriver/IOAHCIBlockStorageDevice, ATA device
    """
    return stdout, stderr


@pytest.fixture
def macos_cmd_capabilities():
    """
    smartctl -c /dev/disk1
    """
    stderr = ""
    stdout = """
smartctl 6.5 2016-05-07 r4318 [Darwin 16.4.0 x86_64] (local build)
Copyright (C) 2002-16, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF READ SMART DATA SECTION ===
General SMART Values:
Offline data collection status:  (0x00)	Offline data collection activity
					was never started.
					Auto Offline Data Collection: Disabled.
Self-test execution status:      (   0)	The previous self-test routine completed
					without error or no self-test has ever
					been run.
Total time to complete Offline
data collection: 		(    0) seconds.
Offline data collection
capabilities: 			 (0x53) SMART execute Offline immediate.
					Auto Offline data collection on/off support.
					Suspend Offline collection upon new
					command.
					No Offline surface scan supported.
					Self-test supported.
					No Conveyance Self-test supported.
					Selective Self-test supported.
SMART capabilities:            (0x0003)	Saves SMART data before entering
					power-saving mode.
					Supports SMART auto save timer.
Error logging capability:        (0x01)	Error logging supported.
					General Purpose Logging supported.
Short self-test routine
recommended polling time: 	 (   2) minutes.
Extended self-test routine
recommended polling time: 	 (  10) minutes.

     """
    return stdout, stderr


class TestDeviceOnMacOs:

    def test_macos(
        self,
        macos_cmd_all,
        macos_cmd_scan_open_as_root,
        macos_cmd_capabilities,
    ):
        """
        This test is to ensure that the parse can be executed without exception.
        """
        from pySMART.device import Device
        with nested(
            mock.patch.object(
                Device,
                "_cmd_all_with_type",
            ),
            mock.patch.object(
                Device,
                "_cmd_scan_open",
            ),
            mock.patch.object(
                Device,
                "_cmd_all",
            ),
            mock.patch.object(
                Device,
                "_cmd_get_capabilities",
            ),
        ) as (mocked_all_with_type, mocked_scan, mocked_all, mocked_cap):
            mocked_all_with_type.return_value = macos_cmd_all
            mocked_scan.return_value = macos_cmd_scan_open_as_root
            mocked_all.return_value = macos_cmd_all
            mocked_cap.return_value = macos_cmd_capabilities
            device = Device("IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/RP06@1C,5/IOPP/SSD0@0/AppleAHCI/PRT0@0/IOAHCIDevice@0/AppleAHCIDiskDriver/IOAHCIBlockStorageDevice", interface='ata')

            assert device.serial == "S29ANYAG123456"
            assert device.model == "APPLE SSD SM0512G"
            assert device.capacity == "500 GB"
            assert device.firmware == "BXW1SA0Q"
            assert device.supports_smart is True
            assert device.messages is not None
            assert device.is_ssd is True
            assert device.assessment == "PASS"
            assert device.get_current_test_status() == (0, "The previous self-test routine completed without error or no self-test has ever been run.")


pySMART/utils.py
#Copyright (C) 2014 Marc Herndon
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License,
#version 2, as published by the Free Software Foundation.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
"""
This module contains generic utilities and configuration information for use
by the other submodules of the `pySMART` package.
"""
# Python built-ins
import ctypes
import os
import platform
from subprocess import Popen, PIPE
from functools import wraps
import warnings

# Configuration definitions
_min_smartctl_ver = {
    'Windows' : {
        'maj' : 6,
        'min' : 1
        },
    'Linux' : {
        'maj' : 5,
        'min' : 42
        },
    'FreeBSD' : {
        'maj' : 6,
        'min' : 1
        },
    'Darwin' : {
        'maj' : 6,
        'min' : 5
        }
    }
"""
(dict of dict of int): Outer dict contains operating system names as keys.
Inner dict has keys 'maj' and 'min' with int values representing the minimum
required major and minor versions, respectively.
"""

OS = platform.system()
"""**(str):** The operating system's name, generally 'Linux' or 'Windows'"""

_req_ma, _req_mi = _min_smartctl_ver[OS]['maj'], _min_smartctl_ver[OS]['min']
"""Major and minor version requirements, parsed from the version string."""

smartctl_type = {
    'ata' : 'ata',
    'csmi' : 'ata',
    'sas' : 'scsi',
    'sat' : 'sat',
    'sata' : 'ata',
    'scsi' : 'scsi'
}
"""
**(dict of str):** Contains actual interface types (ie: sas, csmi) as keys and
the corresponding smartctl interface type (ie: scsi, ata) as values.
"""

# Helper functions
def admin():
    """Determine whether this scrpt is running with administrative privilege.

    ### Returns:
    * **(bool):** True if running as an administrator, False otherwise.
    """
    try:
        is_admin = os.getuid() == 0
    except AttributeError:
        is_admin = ctypes.windll.shell32.IsUserAnAdmin() != 0
    return is_admin

def pd_to_sd(pd):
    """
    Converts a device name from Windows' physical device ID (ie: pd0) to
    Linux's sda notation. Handles up to 'pd675' = 'sdzz'.

    ###Args:
    * **pd (int):** Physical device ID number.

    ##Returns:
    * **(str):** Linux-style 'sd_' device name.
    """
    try:
        pd = int(pd)
    except ValueError:
        return None
    pd2sd = {}
    # Tried to build a list comprehension but Py2.6 on Linux gave syntax error
    for i in range(26):
        pd2sd[i] = chr(ord('a') + i)
    if pd > 26:
        first = (pd // 26) - 1
        second = pd % 26
        return 'sd' + pd2sd[first] + pd2sd[second]
    else:
        return 'sd' + pd2sd[pd]

def rescan_device_busses():
    """Force a rescan of internal storage busses under Windows"""
    cmd = Popen('echo "rescan" | diskpart', shell=True,
                stdout=PIPE, stderr=PIPE)
    _stdout, _stderr = cmd.communicate()

def _warning_on_one_line(message, category, filename, lineno, file=None,
                         line=None):
    """Formats warning messages to appear on one line."""
    return '%s:%s: %s: %s\n' % (filename, lineno, category.__name__, message)
warnings.formatwarning = _warning_on_one_line


def cmd_convert_bytes(func):
    """Keep python 2/3 compatibility by converting Popen bytes into str where
    appropriate."""

    @wraps(func)
    def wrapper(*args, **kwargs):
        stdout, stderr = func(*args, **kwargs)
        if isinstance(stdout, bytes):
            stdout = stdout.decode()
        if isinstance(stderr, bytes):
            stderr = stderr.decode()
        return stdout, stderr
    return wrapper


@cmd_convert_bytes
def cmd_smartctl_version():
    cmd = Popen('smartctl --version', shell=True, stdout=PIPE, stderr=PIPE)
    return cmd.communicate()


def verify_smartctl():
    # Verify smartctl is on the system path and meets the minimum required version
    _stdout, _stderr = cmd_smartctl_version()
    if _stdout == '':
        raise Exception(
            "Required package 'smartmontools' is not installed, or 'smartctl'\n"
            "component is not on the system path. Please install and try again.")
    else:
        for line in _stdout.split('\n'):
            if 'release' in line:
                _ma, _mi = line.strip().split(' ')[2].split('.')
                if (int(_ma) < _req_ma or
                        (int(_ma) == _req_ma and int(_mi) < _req_mi)):
                    raise Exception(
                        "Installed version of smartctl [{0}.{1}] is below the "
                        "minimum requirement of [{2}.{3}]. Please upgrade and "
                        "try again.".format(_ma, _mi, _req_ma, _req_mi))

    # Check for admin rights
    if not admin():
        warnings.warn(
            "_NOT_ADMIN_: smartctl is intended to be run as administrator/root "
            "and may not detect all device types, or may parse device information "
            "incorrectly, if run without these permissions.")

__all__ = ['admin', 'cmd_convert_bytes', 'OS', 'pd_to_sd', 'rescan_device_busses', 'smartctl_type']


pySMART/attribute.py
#Copyright (C) 2014 Marc Herndon
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License,
#version 2, as published by the Free Software Foundation.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
"""
This module contains the definition of the `Attribute` class, used to represent
individual SMART attributes associated with a `Device`.
"""
class Attribute(object):
    """
    Contains all of the information associated with a single SMART attribute
    in a `Device`'s SMART table. This data is intended to exactly mirror that
    obtained through smartctl.
    """
    def __init__(self, num, name, flags, value, worst, thresh, attr_type,
                 updated, when_failed, raw):
        self.num = num
        """**(str):** Attribute's ID as a decimal value (1-255)."""
        self.name = name
        """
        **(str):** Attribute's name, as reported by smartmontools' drive.db.
        """
        self.flags = flags
        """**(str):** Attribute flags as a hexadecimal value (ie: 0x0032)."""
        self.value = value
        """**(str):** Attribute's current normalized value."""
        self.worst = worst
        """**(str):** Worst recorded normalized value for this attribute."""
        self.thresh = thresh
        """**(str):** Attribute's failure threshold."""
        self.type = attr_type
        """**(str):** Attribute's type, generally 'pre-fail' or 'old-age'."""
        self.updated = updated
        """
        **(str):** When is this attribute updated? Generally 'Always' or
        'Offline'
        """
        self.when_failed = when_failed
        """
        **(str):** When did this attribute cross below
        `pySMART.attribute.Attribute.thresh`? Reads '-' when not failed.
        Generally either 'FAILING_NOW' or 'In_the_Past' otherwise.
        """
        self.raw = raw
        """**(str):** Attribute's current raw (non-normalized) value."""

    def __repr__(self):
        """Define a basic representation of the class object."""
        return "<SMART Attribute %r %s/%s raw:%s>" % (
            self.name, self.value, self.thresh, self.raw)

    def __str__(self):
        """
        Define a formatted string representation of the object's content.
        In the interest of not overflowing 80-character lines this does not
        print the value of `pySMART.attribute.Attribute.flags_hex`.
        """
        return "{0:>3} {1:24}{2:4}{3:4}{4:4}{5:9}{6:8}{7:12}{8}".format(
            self.num, self.name, self.value, self.worst, self.thresh,
            self.type, self.updated, self.when_failed, self.raw)

__all__ = ['Attribute']


pySMART/device_list.py
#Copyright (C) 2014 Marc Herndon
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License,
#version 2, as published by the Free Software Foundation.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
"""
This module contains the definition of the `DeviceList` class, used to
represent all physical storage devices connected to the system.
Once initialized, the sole member `devices` will contain a list of `Device`
objects.

This class has no public methods.  All interaction should be through the
`Device` class API.
"""
# Python built-ins
from subprocess import Popen, PIPE

# pySMART module imports
from .device import Device
from .utils import OS, rescan_device_busses


class DeviceList(object):
    """
    Represents a list of all the storage devices connected to this computer.
    """
    def __init__(self, init=True):
        """
        Instantiates and optionally initializes the `DeviceList`.

        ###Args:
        * **init (bool):** By default, `pySMART.device_list.DeviceList.devices`
        is populated with `Device` objects during instantiation. Setting init to
        False will skip initialization and create an empty
        `pySMART.device_list.DeviceList` object instead.
        """
        self.devices = []
        """
        **(list of `Device`):** Contains all storage devices detected during
        instantiation, as `Device` objects.
        """
        if init:
            self._initialize()

    def __repr__(self):
        """Define a basic representation of the class object."""
        rep = "<DeviceList contents:\n"
        for device in self.devices:
            rep += str(device) + '\n'
        return rep + '>'
        #return "<DeviceList contents:%r>" % (self.devices)

    def _cleanup(self):
        """
        Removes duplicate ATA devices that correspond to an existing CSMI
        device. Also removes any device with no capacity value, as this
        indicates removable storage, ie: CD/DVD-ROM, ZIP, etc.
        """
        # We can't operate directly on the list while we're iterating
        # over it, so we collect indeces to delete and remove them later
        to_delete = []
        # Enumerate the list to get tuples containing indeces and values
        for index, device in enumerate(self.devices):
            if device.interface == 'csmi':
                for otherindex, otherdevice in enumerate(self.devices):
                    if (otherdevice.interface == 'ata' or
                            otherdevice.interface == 'sata'):
                        if device.serial == otherdevice.serial:
                            to_delete.append(otherindex)
                            device._sd_name = otherdevice.name
            if device.capacity is None and index not in to_delete:
                to_delete.append(index)
        # Recreate the self.devices list without the marked indeces
        self.devices[:] = [v for i, v in enumerate(self.devices)
                           if i not in to_delete]

    def _initialize(self):
        """
        Scans system busses for attached devices and add them to the
        `DeviceList` as `Device` objects.
        """
        # On Windows machines we should re-initialize the system busses
        # before scanning for disks
        if OS == 'Windows':
            rescan_device_busses()
        cmd = Popen('smartctl --scan-open', shell=True,
                    stdout=PIPE, stderr=PIPE)
        _stdout, _stderr = cmd.communicate()
        for line in _stdout.decode('utf-8').split('\n'):
            if 'failed:' in line or line == '':
                continue

            line_parts = line.split(' ')
            name = line_parts[0]
            # By default device types will be disambiguated by Device.__init__
            interface = None

            if name[0:4] == 'csmi':
                # CSMI devices are explicitly of the 'csmi' type and do not
                # require further disambiguation
                interface = 'csmi'
            elif line_parts[1] == '-d' and ',' in line_parts[2]:
                # More complex cases of device type, see smartctl manual for
                # the details. Do not require further disambiguation
                interface = line_parts[2]
            self.devices.append(Device(name, interface=interface))

        # Remove duplicates and unwanted devices (optical, etc.) from the list
        self._cleanup()
        # Sort the list alphabetically by device name
        self.devices.sort(key=lambda device: device.name)


__all__ = ['DeviceList']


pySMART/__init__.py
#Copyright (C) 2014 Marc Herndon
#
#This program is free software; you can redistribute it and/or
#modify it under the terms of the GNU General Public License,
#version 2, as published by the Free Software Foundation.
#
#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#
#You should have received a copy of the GNU General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
"""
Copyright (C) 2014 Marc Herndon

pySMART is a simple Python wrapper for the `smartctl` component of
`smartmontools`. It works under Linux and Windows, as long as smartctl is on
the system path. Running with administrative (root) privilege is strongly
recommended, as smartctl cannot accurately detect all device types or parse
all SMART information without full permissions.

With only a device's name (ie: /dev/sda, pd0), the API will create a
`Device` object, populated with all relevant information about
that device. The documented API can then be used to query this object for
information, initiate device self-tests, and perform other functions.

Usage
-----
The most common way to use pySMART is to create a logical representation of the
physical storage device that you would like to work with, as shown:

    #!bash
    >>> from pySMART import Device
    >>> sda = Device('/dev/sda')
    >>> sda
    <SATA device on /dev/sda mod:WDC WD5000AAKS-60Z1A0 sn:WD-WCAWFxxxxxxx>

`Device` class members can be accessed directly, and a number of helper methods
are provided to retrieve information in bulk.  Some examples are shown below:

    #!bash
    >>> sda.assessment  # Query the SMART self-assessment
    'PASS'
    >>> sda.attributes[9]  # Query a single SMART attribute
    <SMART Attribute 'Power_On_Hours' 068/000 raw:23644>
    >>> sda.all_attributes()  # Print the entire SMART attribute table
    ID# ATTRIBUTE_NAME          CUR WST THR TYPE     UPDATED WHEN_FAIL    RAW
      1 Raw_Read_Error_Rate     200 200 051 Pre-fail Always  -           0
      3 Spin_Up_Time            141 140 021 Pre-fail Always  -           3908
      4 Start_Stop_Count        098 098 000 Old_age  Always  -           2690
      5 Reallocated_Sector_Ct   200 200 140 Pre-fail Always  -           0
        ... # Edited for brevity
    199 UDMA_CRC_Error_Count    200 200 000 Old_age  Always  -           0
    200 Multi_Zone_Error_Rate   200 200 000 Old_age  Offline -           0
    >>> sda.tests[0]  # Query the most recent self-test result
    <SMART Self-test [Short offline|Completed without error] hrs:23734 LBA:->
    >>> sda.all_selftests()  # Print the entire self-test log
    ID Test_Description Status                        Left Hours  1st_Error@LBA
     1 Short offline    Completed without error       00%  23734  -
     2 Short offline    Completed without error       00%  23734  -
       ... # Edited for brevity
     7 Short offline    Completed without error       00%  23726  -
     8 Short offline    Completed without error       00%  1      -

Alternatively, the package provides a `DeviceList` class. When instantiated,
this will auto-detect all local storage devices and create a list containing
one `Device` object for each detected storage device.

    #!bash
    >>> from pySMART import DeviceList
    >>> devlist = DeviceList()
    >>> devlist
    <DeviceList contents:
    <SAT device on /dev/sdb mod:WDC WD20EADS-00R6B0 sn:WD-WCAVYxxxxxxx>
    <SAT device on /dev/sdc mod:WDC WD20EADS-00S2B0 sn:WD-WCAVYxxxxxxx>
    <CSMI device on /dev/csmi0,0 mod:WDC WD5000AAKS-60Z1A0 sn:WD-WCAWFxxxxxxx>
    >
    >>> devlist.devices[0].attributes[5]  # Access Device data as above
    <SMART Attribute 'Reallocated_Sector_Ct' 173/140 raw:214>

Using the pySMART wrapper, Python applications be be rapidly developed to take
advantage of the powerful features of smartmontools.

Acknowledgements
----------------
I would like to thank the entire team behind smartmontools for creating and
maintaining such a fantastic product.

In particular I want to thank Christian Franke, who maintains the Windows port
of the software.  For several years I have written Windows batch files that
rely on smartctl.exe to automate evaluation and testing of large pools of
storage devices under Windows.  Without his work, my job would have been
significantly more miserable. :)

Having recently migrated my development from Batch to Python for Linux
portabiity, I thought a simple wrapper for smartctl would save time in the
development of future automated test tools.
"""
from .attribute import Attribute
from .device import Device
from .device_list import DeviceList
from .test_entry import Test_Entry
from . import utils

__version__ = '0.3.9'

